voter/src/main/java/org/terracotta/voter/ActiveVoter.java

```diff
@@ -316,27 +339,93 @@ public void registerAndHeartbeat(ExecutorService executorService, ClientVoterMan
 //  by here, the server being voted for seems to not have won the election.  
 //  reset everything and start over
     voteOwner.set(null);
-    futures.forEach(f->f.cancel(true));
+    heartbeatThreads.forEach((hostPort,heartbeatThread) -> heartbeatThread.interrupt());
+    heartbeatThreads.clear();
     voterManagers.stream().forEach(ClientVoterManager::close);
     LOGGER.warn("Rejected by all servers. Attempting to re-register");
   }
-  
+
+  private void pollForTopologyFromActive(Optional<Properties> connectionProps,
+                                         AtomicReference<ClientVoterManager> voteOwner) {
+    heartbeatThreads.put("polling", new Thread(() -> {
+      while (!Thread.currentThread().isInterrupted()) {
+        try {
+          activeClientVoterManager.connect(connectionProps);
+          while (activeClientVoterManager.isConnected()) {
+            String newTopology = activeClientVoterManager.getTopology();
+            LOGGER.info("Topology is {}.", newTopology);
+            if (!existingTopology.equals(newTopology)) {
+              LOGGER.info("New Topology detected {}.", newTopology);
+              // Start heratbeating with new servers
+              List<String> addedServers = getNewlyAddedServer(newTopology);
+              addedServers.forEach(server -> {
+                ClientVoterManager clientVoterManager = new ClientVoterManagerImpl(server);
+                voterManagers.add(clientVoterManager);
+                synchronized (registrationLatch) {
+                  registrationLatch.add(server);
+                }
+                heartbeat(clientVoterManager, connectionProps, voteOwner);
+                heartbeatThreads.get(server).start();
+              });
+
+              // Do removal of old servers from topology
+              List<String> removedServers = getRemovedServer(newTopology);
+              removedServers.forEach(server -> {
+                heartbeatThreads.get(server).interrupt();
+                heartbeatThreads.remove(server);
+              });
+              existingTopology = newTopology;
+            }
+            sleepFor10();
+          }
+        } catch (TimeoutException | RuntimeException e) {
+          activeClientVoterManager.close();
+          sleepFor10();
+        }
+      }
+    }));
+  }
+
+  private List<String> getRemovedServer(String newTopology) {
+    List<String> res = new ArrayList<>();
+    String hostPorts[] = existingTopology.split(",");
+    for (int i = 0; i < hostPorts.length; ++i) {
+      String tmpHostPort = hostPorts[i];
+      if (!newTopology.contains(tmpHostPort)) {
+        res.add(tmpHostPort);
+      }
+    }
+    return res;
+  }
+
+  private List<String> getNewlyAddedServer(String newTopology) {
+    List<String> res = new ArrayList<>();
+    String hostPorts[] = newTopology.split(",");
+    for (int i = 0; i < hostPorts.length; ++i) {
+      String tmpHostPort = hostPorts[i];
+      if (!existingTopology.contains(tmpHostPort)) {
```

Comment:
this code fails with below topology changes

`existingTopology` -> `localhost:12345`
`newTopology` -> `localhost:12345,localhost:1234`